// @datadog/openfeature-node-server/esm/initialization-controller.js
/**
 * Manages deferred initialization with timeout support.
 * Encapsulates the pattern of waiting for external resolution/rejection
 * with automatic timeout handling and cleanup.
 */
export class InitializationController {
    constructor(timeoutMs, onTimeout) {
        this.promise = Promise.race([
            // Deferred promise - resolved/rejected externally
            new Promise((resolve, reject) => {
                this.resolve = () => {
                    this.cleanup();
                    resolve();
                };
                this.reject = (reason) => {
                    this.cleanup();
                    reject(reason);
                };
            }),
            // Timeout promise
            new Promise(() => {
                this.timeoutId = setTimeout(() => {
                    onTimeout();
                }, timeoutMs);
            }),
        ]);
    }
    /**
     * Wait for initialization to complete (or timeout)
     */
    async wait() {
        try {
            await this.promise;
        }
        finally {
            this.cleanup();
        }
    }
    /**
     * Complete initialization successfully
     */
    complete() {
        if (this.resolve) {
            this.resolve();
        }
    }
    /**
     * Fail initialization with an error
     */
    fail(error) {
        if (this.reject) {
            this.reject(error);
        }
    }
    /**
     * Check if initialization is still in progress
     */
    isInitializing() {
        return this.resolve !== undefined;
    }
    cleanup() {
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = undefined;
        }
        this.resolve = undefined;
        this.reject = undefined;
    }
}
