// @datadog/flagging-core/cjs/configuration/flagEvaluationEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFlagEvaluationEvent = createFlagEvaluationEvent;
function createFlagEvaluationEvent(data, timestamp) {
    const event = {
        flag: {
            key: data.flagKey,
        },
        first_evaluation: data.firstEvaluation,
        last_evaluation: data.lastEvaluation,
        evaluation_count: data.count,
        runtime_default_used: data.runtimeDefaultUsed,
        timestamp,
    };
    if (data.targetingKey != null) {
        event.targeting_key = data.targetingKey;
    }
    if (data.error) {
        event.error = { message: data.error };
    }
    if (data.variantKey) {
        event.variant = { key: data.variantKey };
    }
    if (data.allocationKey) {
        event.allocation = { key: data.allocationKey };
    }
    if (data.targetingRuleKey) {
        event.targeting_rule = { key: data.targetingRuleKey };
    }
    // Add context data if targeting context exists
    if (data.targetingContext && Object.keys(data.targetingContext).length > 0) {
        event.context = {
            evaluation: data.targetingContext,
        };
    }
    return event;
}
