// @datadog/flagging-core/cjs/configuration/flagEvaluationAggregator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlagEvaluationAggregator = void 0;
const obfuscation_1 = require("../obfuscation");
const flagEvaluationEvent_1 = require("./flagEvaluationEvent");
class FlagEvaluationAggregator {
    constructor(flushInterval, onFlush) {
        this.aggregatedData = new Map();
        this.flushInterval = flushInterval;
        this.onFlush = onFlush;
    }
    start() {
        this.intervalId = setInterval(() => {
            this.flush();
        }, this.flushInterval);
    }
    stop() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = undefined;
        }
        this.flush();
    }
    addEvaluation(context, details, error) {
        var _a, _b;
        const keyString = this.createAggregationKeyString(context, details, error);
        const timestamp = Date.now();
        const existingData = this.aggregatedData.get(keyString);
        if (existingData) {
            existingData.count++;
            existingData.lastEvaluation = timestamp;
            if (error) {
                existingData.error = error;
            }
        }
        else {
            const runtimeDefaultUsed = details.reason === 'DEFAULT' || details.reason === 'ERROR';
            const allocationKey = (_a = details.flagMetadata) === null || _a === void 0 ? void 0 : _a.allocationKey;
            const targetingRuleKey = (_b = details.flagMetadata) === null || _b === void 0 ? void 0 : _b.targetingRuleKey;
            const { targetingKey, ...targetingContext } = context;
            this.aggregatedData.set(keyString, {
                flagKey: details.flagKey,
                variantKey: details.variant,
                allocationKey,
                targetingRuleKey,
                targetingKey,
                targetingContext,
                count: 1,
                firstEvaluation: timestamp,
                lastEvaluation: timestamp,
                runtimeDefaultUsed,
                error,
            });
        }
    }
    flush() {
        if (this.aggregatedData.size === 0) {
            return;
        }
        const events = Array.from(this.aggregatedData.values()).map((data) => (0, flagEvaluationEvent_1.createFlagEvaluationEvent)(data, data.firstEvaluation));
        this.aggregatedData.clear();
        this.onFlush(events);
    }
    createAggregationKeyString(context, details, error) {
        var _a, _b;
        const allocationKey = (_a = details.flagMetadata) === null || _a === void 0 ? void 0 : _a.allocationKey;
        const targetingRuleKey = (_b = details.flagMetadata) === null || _b === void 0 ? void 0 : _b.targetingRuleKey;
        const { targetingKey, ...targetingContext } = context;
        // Hash the deterministic object representation
        return (0, obfuscation_1.getMD5Hash)(JSON.stringify({
            flagKey: details.flagKey,
            variant: details.variant || '',
            allocationKey: allocationKey || '',
            targetingRuleKey: targetingRuleKey || '',
            targetingKey: targetingKey || '',
            targetingContext,
            error: error || '',
        }));
    }
}
exports.FlagEvaluationAggregator = FlagEvaluationAggregator;
